src/Microsoft.AspNet.MiddlewareAnalysis/AnalysisBuilder.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http;
+using Microsoft.AspNet.Http.Features;
+
+namespace Microsoft.AspNet.MiddlewareAnalysis
+{
+    public class AnalysisBuilder : IApplicationBuilder
+    {
+        private const string NextMiddlewareName = "analysis.NextMiddlewareName";
+
+        public AnalysisBuilder(IApplicationBuilder inner)
+        {
+            InnerBuilder = inner;
+        }
+
+        private IApplicationBuilder InnerBuilder { get; }
+
+        public IServiceProvider ApplicationServices
+        {
+            get { return InnerBuilder.ApplicationServices; }
+            set { InnerBuilder.ApplicationServices = value; }
+        }
+
+        public IDictionary<string, object> Properties
+        {
+            get { return InnerBuilder.Properties; }
+        }
+
+        public IFeatureCollection ServerFeatures
+        {
+            get { return InnerBuilder.ServerFeatures;}
+        }
+
+        public RequestDelegate Build()
+        {
+            // Add one maker at the end before the default 404 middleware (or any fancy Join middleware).
+            return InnerBuilder.UseMiddleware<AnalysisMiddleware>("EndOfPipeline")
```

Comment:
This Build is also only ever called on branches that have first called New (e.g. Map). See https://github.com/aspnet/Diagnostics/pull/207/files#diff-9ac90184c9cfc4a15d6622dbce5b5cf6R19